rust/src/client/mobile_client/client.rs

```diff
@@ -0,0 +1,364 @@
+use crate::{
+    client::{
+        mobile_client::participant::{
+            Awaiting,
+            Participant,
+            ParticipantSettings,
+            Role,
+            Sum,
+            Sum2,
+            Update,
+        },
+        ClientError,
+        Proxy,
+    },
+    crypto::ByteObject,
+    mask::model::Model,
+    state_machine::coordinator::RoundParameters,
+    InitError,
+    PetError,
+};
+use derive_more::From;
+use std::{cell::RefCell, rc::Rc};
+
+#[derive(From)]
+pub enum ClientStateMachine {
+    Awaiting(ClientState<Awaiting>),
+    Sum(ClientState<Sum>),
+    Update(ClientState<Update>),
+    Sum2(ClientState<Sum2>),
+}
+
+impl ClientStateMachine {
+    pub fn new(
+        proxy: Proxy,
+        participant_settings: ParticipantSettings,
+        local_model: Rc<RefCell<Option<Model>>>,
+        global_model: Rc<RefCell<Option<Model>>>,
+    ) -> Result<Self, InitError> {
+        // crucial: init must be called before anything else in this module
+        sodiumoxide::init().or(Err(InitError))?;
+
+        Ok(ClientState::<Awaiting>::new(
+            proxy,
+            Participant::<Awaiting>::new(participant_settings.into()),
+            local_model,
+            global_model,
+        )
+        .into())
+    }
+
+    pub async fn next(self) -> Self {
+        match self {
+            ClientStateMachine::Awaiting(state) => state.next().await,
+            ClientStateMachine::Sum(state) => state.next().await,
+            ClientStateMachine::Update(state) => state.next().await,
+            ClientStateMachine::Sum2(state) => state.next().await,
+        }
+    }
+}
+
+pub struct ClientState<Type> {
+    proxy: Proxy,
+    round_params: RoundParameters,
+    participant: Participant<Type>,
+    local_model: Rc<RefCell<Option<Model>>>,
```

Comment:
Why do we need these to be wrapped into `Rc`? In `run()` and `fetch_global_model()`, where these two fields are used, it seems that we could do without it. Removing the `Rc<RefCell>` would be nice if we want to share this code with the "regular" multi-threaded client, because `ClientState` would be `Send`.

Reply:
That is a good point! Do you have something in mind how we could solve this?
Two solutions that I can think of right now are:

1. Replace `local`/`global` model with a separate component in the `ClientState` that is responsible for `get_local_model` and `set_global_model`. We might even define a trait here and have different implementations for each environment (web, mobile, ...). 
For example:
for the mobile environment we might need to store it on the file system but on a desktop environment it might be ok to keep it in memory. 
 
2. We just add the local/global model to the client state.

Reply:
I think on the long run #1 is the best solution. But this PR is already pretty massive so I think #2 is a good enough first implementation. The way I see it, we can "just" remove the `Rc<RefCell<>>` wrapping and adjust the code wherever rustc complains. But I haven't tried it so I may be missing something.

Reply:
Alright I removed the `Rc<RefCell<>>` and moved the `local` and `global` model into the `ClientState` struct. Maybe not the most elegant solution but I think it is good for now but I'm open to change it to something different.